### PR TITLE
[TF FE] Extended support of tagged MetaGraphs and variables handling in SavedModel

### PR DESCRIPTION
**Pull request by gkrivor**
### Details:
- Added support of alternate MetaGraph tags
- Added support of variables without explicit RestoreV2

### Tickets:
 - 111260, 111262

### CODE REVIEW

File: src/frontends/tensorflow/src/graph_iterator_saved_model.hpp [hunk 1]
```diff
@@ -85,68 +85,97 @@ class GraphIteratorSavedModel : public GraphIteratorProto {
         bool res = m_saved_model->ParseFromIstream(&sm_stream);
         FRONT_END_GENERAL_CHECK(res && m_saved_model->meta_graphs_size(), "Saved Model cannot be parsed");
 
+        const ::tensorflow::MetaGraphDef* meta_graph_with_no_tags = nullptr;
+
```
Comment (rkazants) [inline src/frontends/tensorflow/src/graph_iterator_saved_model.hpp:89]: It looks that `meta_graph_with_no_tags` is not used. Please check

File: src/frontends/tensorflow/src/graph_iterator_saved_model.hpp [hunk 2]
```diff
@@ -85,68 +85,97 @@ class GraphIteratorSavedModel : public GraphIteratorProto {
...
         for (const auto& meta_graph : m_saved_model->meta_graphs()) {
```
Comment (rkazants) [inline src/frontends/tensorflow/src/graph_iterator_saved_model.hpp:90]: minor:
please add a comment above `for`:
`SavedModel can contain several MetaGraph with different tags. Look for MetaGraph with the required tag`

File: src/frontends/tensorflow/src/graph_iterator_saved_model.hpp [hunk 3]
```diff
@@ -85,68 +85,97 @@ class GraphIteratorSavedModel : public GraphIteratorProto {
...
                 }
-                if (!tag_found) {
+            }
+            if (tag_found) {
+                return load_meta_graph(meta_graph);
+            }
+        }
+
+        // Alternate behavior for working with "default tag" to support additional cases for read_model
+        if (tags == "serve") {
```
Comment (rkazants) [inline src/frontends/tensorflow/src/graph_iterator_saved_model.hpp:112]: May be it is better to move "serve" into `checkpoint utils` and declare with `#define`?

File: src/frontends/tensorflow/src/graph_iterator_saved_model.hpp [hunk 4]
```diff
@@ -85,68 +85,97 @@ class GraphIteratorSavedModel : public GraphIteratorProto {
...
+            bool tag_found = false;
+            for (const auto& tag : meta_graph.meta_info_def().tags()) {
+                if (tags.find(tag) != std::string::npos) {
+                    tag_found = true;
+                    break;
                 }
-                if (!tag_found) {
+            }
+            if (tag_found) {
+                return load_meta_graph(meta_graph);
```
Comment (rkazants) [inline src/frontends/tensorflow/src/graph_iterator_saved_model.hpp:107]: Should not we check that this `MetaGraph` has at least one `GraphDef`?

File: src/frontends/tensorflow/src/graph_iterator_saved_model.hpp [hunk 5]
```diff
@@ -85,68 +85,97 @@ class GraphIteratorSavedModel : public GraphIteratorProto {
...
-            if (m_saved_model->meta_graphs_size() > 1) {
-                bool tag_found = false;
-                for (const auto& tag : meta_graph.meta_info_def().tags()) {
-                    if (tags.find(tag) != std::string::npos) {
-                        tag_found = true;
-                        break;
-                    }
+            bool tag_found = false;
+            for (const auto& tag : meta_graph.meta_info_def().tags()) {
+                if (tags.find(tag) != std::string::npos) {
```
Comment (rkazants) [inline src/frontends/tensorflow/src/graph_iterator_saved_model.hpp:101]: Please double check why `tags` is `std::string`. I though it is a vector of strings

File: src/frontends/tensorflow/src/graph_iterator_saved_model.hpp [hunk 6]
```diff
@@ -85,68 +85,97 @@ class GraphIteratorSavedModel : public GraphIteratorProto {
...
+                if (meta_graph.meta_info_def().tags_size() == 0) {
+                    return load_meta_graph(meta_graph);
                 }
             }
 
-            // MetaGraph may have a list of signatures, but at this moment we need information only about
-            // "serving_default" signature which contains information about inputs/outputs names for the
-            // model. Situation when it is missing in a file also could be.
-            auto serving_default = validSignatures.find("serving_default");
+            FRONT_END_GENERAL_CHECK(false, "Saved Model doesn't contain any applicable MetaGraph");
```
Comment (rkazants) [inline src/frontends/tensorflow/src/graph_iterator_saved_model.hpp:130]: please add `[TensorFlow Frontend]`. It will help to identify TF FE error

**Review by rkazants — COMMENTED** [review]
Great test!
Left comments regarding the code

File: src/frontends/tensorflow/src/graph_iterator_saved_model.hpp (hunk 2 above)
Comment (gkrivor) [reply src/frontends/tensorflow/src/graph_iterator_saved_model.hpp:90]: ok

File: src/frontends/tensorflow/src/graph_iterator_saved_model.hpp (hunk 1 above)
Comment (gkrivor) [reply src/frontends/tensorflow/src/graph_iterator_saved_model.hpp:89]: Yep, it is an artifact, removed

File: src/frontends/tensorflow/src/graph_iterator_saved_model.hpp (hunk 5 above)
Comment (gkrivor) [reply src/frontends/tensorflow/src/graph_iterator_saved_model.hpp:101]: Regarding information I found, I'll change this part. We should parse "tags" with ","-delimeter and compare resulting vectors.

File: src/frontends/tensorflow/src/graph_iterator_saved_model.hpp (hunk 4 above)
Comment (gkrivor) [reply src/frontends/tensorflow/src/graph_iterator_saved_model.hpp:107]: Here - no, we skip all MetaGraphs without GraphDef (has_graph_def() == false) at line 89/95 above

File: src/frontends/tensorflow/src/graph_iterator_saved_model.hpp (hunk 3 above)
Comment (gkrivor) [reply src/frontends/tensorflow/src/graph_iterator_saved_model.hpp:112]: done

File: src/frontends/tensorflow/src/graph_iterator_saved_model.hpp (hunk 6 above)
Comment (gkrivor) [reply src/frontends/tensorflow/src/graph_iterator_saved_model.hpp:130]: Np. I strongly suggest to define this part somewhere else and use it as a constant. We have nonzero-overhead adding it each time into all strings.

Comment (rkazants) [reply src/frontends/tensorflow/src/graph_iterator_saved_model.hpp:130]: We have a ticket for this activity. Agree, I also don't like it to add it everytime

File: src/frontends/tensorflow/src/graph_iterator_saved_model.cpp [hunk 7]
```diff
@@ -72,6 +72,31 @@ std::basic_string<wchar_t> get_variables_index_name<wchar_t>() {
 }
 #endif
 
+std::vector<std::string> GraphIteratorSavedModel::split_tags(const std::string tags) const {
```
Comment (rkazants) [inline src/frontends/tensorflow/src/graph_iterator_saved_model.cpp:75]:
```suggestion
std::vector<std::string> GraphIteratorSavedModel::split_tags(const std::string& tags) const {
```

Comment (rkazants) [reply src/frontends/tensorflow/src/graph_iterator_saved_model.cpp:75]: Minor, can be fixed in the next PR

**Review by rkazants — APPROVED** [review]

**Pull request merged**